File: python/2015/day03b/day03b.py
#!/usr/bin/env python3

import sys

def usage():
    progname = sys.argv[0]
    print(f'usage: {progname} <input file>')
    sys.exit(1)

def process(filename):
    positions = set()
    santa = (0, 0)
    roboSanta = (0, 0)
    i = 0
    positions.add(santa)
    with open(filename, 'r') as infile:
        for line in infile:
            for ch in line:
                if i % 2 == 0:
                    if ch == '^':
                        santa = (santa[0], santa[1] + 1)
                    elif ch == 'v':
                        santa = (santa[0], santa[1] - 1)
                    elif ch == '<':
                        santa = (santa[0] - 1, santa[1])
                    elif ch == '>':
                        santa = (santa[0] + 1, santa[1])
                    positions.add(santa)
                else:
                    if ch == '^':
                        roboSanta = (roboSanta[0], roboSanta[1] + 1)
                    elif ch == 'v':
                        roboSanta = (roboSanta[0], roboSanta[1] - 1)
                    elif ch == '<':
                        roboSanta = (roboSanta[0] - 1, roboSanta[1])
                    elif ch == '>':
                        roboSanta = (roboSanta[0] + 1, roboSanta[1])
                    positions.add(roboSanta)
                i += 1
    return len(positions)

def main():
    if len(sys.argv) < 2:
        usage()
    filename = sys.argv[1]
    result = process(filename)
    print(f'result = {result}')

if __name__ == '__main__':
    main()

File: perl/2015/day03b/day03b.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub usage {
    my $progname = $0;
    print "usage: $progname <input file>\n";
    exit 1;
}

sub process {
    my ($filename) = @_;
    my %positions;
    my $santa = [0, 0];
    my $roboSanta = [0, 0];
    my $position = join ',', @$santa;
    $positions{$position} = 1;
    my $i = 0;

    open ( my $infile, '<', $filename ) or die "cannot open file: $filename: $!";
    while ( my $line = <$infile> ) {
        foreach my $ch ( split //, $line ) {
            my $santaMove = $i % 2;
            if ( $ch eq '^' ) {
                $santaMove ? ($santa->[1] += 1) : ($roboSanta->[1] += 1);
            } elsif ( $ch eq 'v' ) {
                $santaMove ? ($santa->[1] -= 1) : ($roboSanta->[1] -= 1);
            } elsif ( $ch eq '<' ) {
                $santaMove ? ($santa->[0] -= 1) : ($roboSanta->[0] -= 1);
            } elsif ( $ch eq '>' ) {
                $santaMove ? ($santa->[0] += 1) : ($roboSanta->[0] += 1);
            }
            $position = join ',', $santaMove ? @$santa : @$roboSanta;
            $positions{$position} = 1;
            $i++;
        }
    }
    close($infile);

    return scalar keys %positions;
}

sub main {
    if ( @ARGV < 1 ) {
        usage();
    }
    my $filename = $ARGV[0];
    my $result = process($filename);
    print "result = $result\n";
}

main();
